// umesse-lib/constants.js
"use strict";

const { randomBytes } = require("crypto");

exports.constants = Object.freeze({
  region: "ap-northeast-1",
  debug: process.env.debug,
  environment: process.env.environment,

  // S3 設定
  s3Config: function () {
    switch (this.environment) {
      case "local":
        return {
          region: this.region,
          endpoint: "http://localhost:4566",
          signatureVersion: "v4",
          s3ForcePathStyle: "true", // local only
        };
      case "localstack":
        return {
          region: this.region,
          endpoint: "http://host.docker.internal:4566",
          signatureVersion: "v4",
          s3ForcePathStyle: "true", // local only
        };
      case "dev":
      case "stg":
      case "prod":
      default:
        return {
          region: this.region,
        };
    }
  },
  // S3 バケット
  s3Bucket: function () {
    switch (this.environment) {
      case "local":
      case "localstack":
        return {
          users: "umesse-users",
          contents: "umesse-contents",
        };
      case "dev":
        return {
          users: "dev-umesse-users",
          contents: "dev-umesse-contents",
        };
      case "stg":
        return {
          users: "stg-umesse-users",
          contents: "stg-umesse-contents",
        };
      case "prod":
      default:
        return {
          users: "umesse-users",
          contents: "umesse-contents",
        };
    }
  },

  // DynamoDB 設定
  dynamoDbConfig: function () {
    switch (this.environment) {
      case "local":
        return {
          region: this.region,
          endpoint: "http://localhost:4566",
        };
      case "localstack":
        return {
          region: this.region,
          endpoint: "http://host.docker.internal:4566",
        };
      case "dev":
      case "stg":
      case "prod":
      default:
        return {
          region: this.region,
        };
    }
  },
  // DynamoDB テーブル
  dynamoDbTable: function () {
    switch (this.environment) {
      case "local":
      case "localstack":
        return {
          users: "umesse-users",
          contents: "umesse-contents",
          external: "umesse-external",
        };
      case "dev":
        return {
          users: "dev-umesse-users",
          contents: "dev-umesse-contents",
          external: "dev-umesse-external",
        };
      case "stg":
        return {
          users: "stg-umesse-users",
          contents: "stg-umesse-contents",
          external: "stg-umesse-external",
        };
      case "prod":
      default:
        return {
          users: "umesse-users",
          contents: "umesse-contents",
          external: "umesse-external",
        };
    }
  },

  // SQS 設定
  sqsConfig: function () {
    switch (this.environment) {
      case "local":
        return {
          region: this.region,
          endpoint: "http://localhost:4566",
        };
      case "localstack":
        return {
          region: this.region,
          endpoint: "http://host.docker.internal:4566",
        };
      case "dev":
      case "stg":
      case "prod":
      default:
        return {
          region: this.region,
        };
    }
  },
  // TODO: SQS QUEUE URL 設定
  sqsQueueUrl: function () {
    switch (this.environment) {
      case "local":
      case "localstack":
        return "http://localhost:4566/000000000000/umesseConverterQueue";
      case "dev":
      case "stg":
      case "prod":
      default:
        return "";
    }
  },

  // TODO: TTS 設定
  ttsConfig: function () {
    switch (this.environment) {
      case "local":
      case "localstack":
      case "dev":
      case "stg":
      case "prod":
      default:
        return {
          host: "api.voicetext.jp",
          path: "/v1/tts",
          key: "yck1tgf1x2jl3m51",
        };
    }
  },

  // CMステータス
  cmStatus: {
    DELETE: "00", // CM削除(ユーザーミスによる復活のための定義)
    CREATING: "01", // CM作成中
    COMPLETE: "02", // CM作成完了
    CONVERT: "03", // CMエンコード中
    SHARING: "04", // CM共有中
    ERROR: "09", // CMエラー
    EXTERNAL_UPLOADING: "11", // 外部システムアップロード中
    EXTERNAL_COMPLETE: "12", // 外部システムアップロード完了
    EXTERNAL_ERROR: "19", // 外部システムアップロードエラー
  },

  // CMタイプ
  cmProductionType: {
    MUSIC: "01", // 音楽系CM
    NONE: "02", // 素ナレCM（BGMなし）
  },

  // CMアップロードシステム区分
  cmUploadSystem: {
    CENTER: "01",
    SSENCE: "02",
  },

  // コンテンツのカテゴリー
  resourceCategory: {
    CM: "cm",
    RECORDING: "recording",
    TTS: "tts",
    BGM: "bgm",
    CHIME: "chime",
    NARRATION: "narration",
  },
});

// debug log
exports.debuglog = (message) => {
  if (this.constants.debug) console.log(`[debug] ${message}`);
};

// error log
exports.errorlog = (message) => {
  console.error(message);
};

// timestamp
exports.timestamp = () => {
  const time = new Date();
  time.setHours(time.getHours() + 9);
  return time.toISOString().split("Z")[0] + "+09:00";
};

// generate id cm (c), recording (r), tts (t)
exports.generateId = (unisCustomerCd, div) => {
  const id = randomBytes(8).reduce((p, i) => p + (i % 36).toString(36), "");
  return `${unisCustomerCd}-${div}-${id}`;
};
