// umesse-lib/converter/index.js
'use strict';


var UMesseConverter = function (unisCustomerCd, cmId) {
    this.ffmpeg = `ffmpeg -hide_banner`;
    this.workDir = `/tmp/${unisCustomerCd}/mix/${cmId}`;
    this.windowsWorkDir = `${process.cwd()}\\tmp\\${unisCustomerCd}\\mix\\${cmId}`; // windows only

    const isWindows = process.platform === 'win32';
    const isLocal = process.env.environment === 'local';
    this.isLocalWindows = isLocal && isWindows;

    // windows用処理, 
    // TODO: third-partyを使用する場合dockerでの処理を変更する必要がある.
    if (this.isLocalWindows)
        require('./win-processor.js')(UMesseConverter.prototype);
    else {
        // TODO: third-partyでいく場合 processorにまとめて書く
        // windowsの処理は、windowsようのffmpegを用意しsetPathFFmpegで設定など行う
        // win-processor.jsは削除できると思う.
        require('./processor.js')(UMesseConverter.prototype);
        require('./third-party')(UMesseConverter.prototype);
    }


}

UMesseConverter.prototype.setWorkDir = function (workDir) {
    this.workDir = workDir;
}
UMesseConverter.prototype.getWorkDir = function () {
    return this.workDir;
}
module.exports = UMesseConverter
